### ISSUE TITLE
Show album art stored in individual songs (enhancement request)

### ISSUE THREAD

**Issue by bastianilso** [vanilla-music/vanilla#713]
I have a lot of music which despite being listed in the same album, have unique individual cover arts embedded in them. It would be really nice if Vanilla music would give priority to show these individual covers in the Now playing screen.

**Comment by adrian-bl** [comment]
The main issue with this is (besides that we require about 20x more storage for covers and CPU time): How do we get the individual covers?

We currently pick up image files located in the same folder OR query androids media database. However: Android only has a single cover per album (= acts the same as we do) and AFAIK there is no way to extract embedded covers using the media framework on Android.

**Comment by bastianilso** [comment]
Good question.. Somehow, my HUAWEI Music system app does show individual song covers, but I don't know if they are using the Android media framework..

**Comment by abn-volk** [comment]
Actually, Androud's MediaMetadataRetriever has a getEmbeddedBitmap() method that can do this: https://developer.android.com/reference/android/media/MediaMetadataRetriever.html#getEmbeddedPicture()

So I think it is completely possible. All that is needed is setting the correct data source and scale large bitmaps manually to prevent OOMs: https://developer.android.com/topic/performance/graphics/load-bitmap.html

**Comment by adrian-bl** [comment]
abn-volk: oh yeah, you are right. I must have missed that last time i looked at it.

So i think we could have this as an option to activate: It will have a significant impact on the initial cover loading (and the amount of covers we can cache), so i wouldn't want to have it enabled by default.

**Comment by Adonai** [comment]
@adrian-bl sounds good to me, > 5.x devices are quite capable for this task anyway

**Comment by cosmicdan** [comment]
Has any preliminary work been done on this issue? Love Vanilla Music as a replacement for the obsolete PowerAMP, but the "wrong" album art appearing for a whole lot of my tracks is a little bit annoying 😀

**Comment by adrian-bl** [comment]
I dont think that anyone took a stab at this so far.

But the implementation should not be too hard.

**Comment by cosmicdan** [comment]
> But the implementation should not be too hard.

Indeed, I might take a crack at it (I have holidays coming up) though I'm a bit rusty with Android-specific Java. I'll do a PR if I get it working.
